diffrax/_adjoint.py

```
@@ -852,3 +852,196 @@ def loop(
+        **kwargs,
+    )
+
+
+@_loop_reversible.def_fwd
+def _loop_reversible_fwd(perturbed, y__args__terms, **kwargs):
+    del perturbed
+    final_state, aux_stats = _loop_reversible(y__args__terms, **kwargs)
+    t1 = final_state.save_state.ts[-1]
+    y1 = final_state.save_state.ys[-1]
+    solver_state1 = final_state.solver_state
+    return (final_state, aux_stats), (t1, y1, solver_state1)
+
+
+@_loop_reversible.def_bwd
+def _loop_reversible_bwd(
+    residuals,
+    grad_final_state__aux_stats,
+    perturbed,
+    y__args__terms,
+    *,
+    self,
+    solver,
+    event,
+    t0,
+    t1,
+    dt0,
+    init_state,
+    progress_meter,
+    **kwargs,
+):
+    assert event is None
+
+    del perturbed, init_state, t1, progress_meter, self, kwargs
+    t1, y1, solver_state1 = residuals
+    original_solver_state, z1 = solver_state1
+    del residuals, solver_state1
+
+    grad_final_state, _ = grad_final_state__aux_stats
+    # ReversibleAdjoint currently only allows SaveAt(t1=True) so grad_y1 should have
+    # the same structure as y1.
+    grad_y1 = grad_final_state.save_state.ys[-1]
+    del grad_final_state, grad_final_state__aux_stats
+
+    y, args, terms = y__args__terms
+    del y__args__terms
+
+    diff_args = eqx.filter(args, eqx.is_inexact_array)
+    diff_terms = eqx.filter(terms, eqx.is_inexact_array)
+    diff_z1 = eqx.filter(z1, eqx.is_inexact_array)
+    grad_args = jtu.tree_map(jnp.zeros_like, diff_args)
+    grad_terms = jtu.tree_map(jnp.zeros_like, diff_terms)
+    grad_z1 = jtu.tree_map(jnp.zeros_like, diff_z1)
+    del diff_args, diff_terms, diff_z1
+
+    def grad_step(state):
+        def solver_step(terms, t0, t1, y1, args):
+            step, _, _, _, _ = solver.solver.step(
+                terms, t0, t1, y1, args, (first_step, f0), False
+            )
+            return step
+
+        t1, y1, solver_state, grad_y1, grad_z1, grad_args, grad_terms = state
+        (first_step, f0), z1 = solver_state
```

So here you are assuming the use of `diffrax.Reversible`, right? You should throw an exception in `ReversibleAdjoint.loop` if that solver is not used, otherwise the user might get a useless "not enough values to unpack" error. (Maybe you raise this somewhere I haven't seen?)